File: apps/api/app/incomes/application/update_income.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from typing import Any, TypeVar

from app.incomes.domain.entities import Income
from app.incomes.domain.errors import IncomeNotFoundError, UnknownSourceError
from app.incomes.ports.repository import IncomeRepository
from app.incomes.ports.source_checker import SourceChecker
from app.shared.domain.money import Money

_T = TypeVar("_T")


@dataclass(frozen=True, slots=True)
class UpdateIncomeCommand:
    income_id: int
    fields: frozenset[str]  # which attributes the caller intends to change
    amount_cents: int | None = None
    currency: str | None = None
    source_id: int | None = None
    occurred_on: date | None = None
    note: str | None = None


class UpdateIncome:
    def __init__(self, repository: IncomeRepository, sources: SourceChecker) -> None:
        self._repository = repository
        self._sources = sources

    def __call__(self, command: UpdateIncomeCommand) -> Income:
        existing = self._repository.get(command.income_id)
        if existing is None:
            raise IncomeNotFoundError(command.income_id)

        amount_cents = self._pick(command, "amount_cents", existing.money.amount_cents)
        currency = self._pick(command, "currency", existing.money.currency)
        source_id = self._pick(command, "source_id", existing.source_id)
        occurred_on = self._pick(command, "occurred_on", existing.occurred_on)
        note = self._pick(command, "note", existing.note)

        if "source_id" in command.fields and not self._sources.exists(source_id):
            raise UnknownSourceError(source_id)

        updated = Income(
            id=existing.id,
            money=Money(amount_cents, currency),
            source_id=source_id,
            occurred_on=occurred_on,
            note=note,
        )
        return self._repository.update(updated)

    @staticmethod
    def _pick(command: UpdateIncomeCommand, name: str, current: _T) -> _T:
        value: Any = getattr(command, name)
        return value if name in command.fields else current

File: apps/api/internal/incomes/application/update_income.go
package application

import (
	"context"

	"caudal-api/internal/incomes/domain"
	"caudal-api/internal/incomes/ports"
	shared "caudal-api/internal/shared/domain"
)

type UpdateIncome struct {
	repository ports.Repository
	checker    ports.SourceChecker
}

func NewUpdateIncome(repository ports.Repository, checker ports.SourceChecker) UpdateIncome {
	return UpdateIncome{repository: repository, checker: checker}
}

func (useCase UpdateIncome) Execute(ctx context.Context, command UpdateIncomeCommand) (domain.Income, error) {
	income, err := useCase.repository.Get(ctx, command.ID)
	if err != nil {
		return domain.Income{}, err
	}
	if err := useCase.applyChanges(ctx, &income, command); err != nil {
		return domain.Income{}, err
	}
	return useCase.repository.Update(ctx, income)
}

func (useCase UpdateIncome) applyChanges(ctx context.Context, income *domain.Income, command UpdateIncomeCommand) error {
	if command.SourceID != nil {
		exists, err := useCase.checker.IncomeSourceExists(ctx, *command.SourceID)
		if err != nil || !exists {
			if err != nil {
				return err
			}
			return domain.ErrUnknownSource
		}
		income.SourceID = *command.SourceID
	}
	if command.AmountCents != nil || command.Currency != nil {
		amount, currency := income.Money.AmountCents, income.Money.Currency
		if command.AmountCents != nil {
			amount = *command.AmountCents
		}
		if command.Currency != nil {
			currency = *command.Currency
		}
		money, err := shared.NewMoney(amount, currency)
		if err != nil {
			return err
		}
		income.Money = money
	}
	if command.OccurredOn != nil {
		income.OccurredOn = *command.OccurredOn
	}
	if command.Note != nil {
		income.Note = *command.Note
	}
	if command.RecurrenceID != nil {
		income.RecurrenceID = *command.RecurrenceID
	}
	return nil
}
